src/core/context.py

```diff
@@ -64,6 +68,12 @@ def _zookeeper_relation(self) -> Relation | None:
         """The zookeeper relation."""
         return self.model.get_relation(ZOOKEEPER_REL)
 
+    @property
+    def _service_manager(self) -> ServiceManager | None:
```

Comment:
**thought, suggestion** I find this rather weird and strange, since we are injecting business logic (since it is a manger) into the section (`core`) where we are only dealing with the context, and state (parsing databags and creating domain objects). 

This breaks a bit the dependency structure, where ideally core is rather dependency free, managers depends on core, events depends on managers (and therefore cores). Here we have core that depends on managers, therefore the circular dependency. 

I have a couple of possible suggestions/solutions:
1. Move the service managers to manages and refactor the code accordingly
2. Split the parsing logic of the ServiceManager into the parsing part, and the business-logic part (like creating the service, reconciling, etc)

I understand that as the current code stands, it may seem simpler to just have a manager here. But if we start to deviate from the pattern here, we may well do it elsewhere, breaking the separation of concerns that we currently have, and ending up with more entropy and less maintanable codebases IMHO. I'm curious what others think. I have placed the comment into `suggestion` since it would not be my call, but it is rather for you guys to make a call (and eventually @welpaolo's being the tech lead for the spark project to ).

Reply:
Alternatively, we could make `kyuubi_address` a method with a `service_manager` parameter. IMO it would address some of your concerns:
- We would avoid instantiating a `ServiceManager` in this module, which is indeed a circular dependency
- Having such a parameter for `kyuubi_address` would only add `ServiceManager` as a type annotation, which would not be a runtime circular dependency (by using `typing.TYPE_CHECKING` to conditionally import it).
- Moreover, using this dependency injection pattern would make this specific function easier to test because we would be able to pass our own `ServiceManager` potentially mocked in the unit tests
- It's more of a nitpick, and a feeble argument, but I usually don't like having an external call behind a non-cached property. Properties usually convey some kind of fast, attribute-like access, whereas we can expect a method call to take some time. It's an even weaker argument because I did exactly that for ZK where we have a property masking an API call, but well, it is what it is

Reply:
I agree here, we need to remove this manager from the context. I will put this logic in the manager. I would like to continue to follow the pattern. @theoctober19th what do you think here?

Reply:
Not sure I like the pattern proposed by @Batalex if I have to be honest. Seems to me a workaround/patch, to avoid having a runtime error, but still having a similar scope cripple and tension between concerns IMHO. I honestly still see the `typing.TYPE_CHECKING` as a python workaround which I would not honestly encourage too much. 

But I agree on his fourth point about using `cached_property` when using an API call. It does not cost much (just the import of funtools decorator), and it well worth the effort and the pattern in general.

Reply:
Thanks everyone for the suggestions, this has been fixed in https://github.com/canonical/kyuubi-k8s-operator/pull/46/commits/67194342b9bae2e88676195b6e0f7e93bd642bc0